Sprint 1/step 2

yuriydyuster opened the pull request:
Исправлены замечания от ревьюера:

- пул-реквест должен быть создан из ветки sprint-1/step-2 в main
- на данном этапе нужно выбрать один из двух подходов
- либо переименовать все компоненты в jsx, оставить PropTypes и убрать все типы typescript
- либо оставить все на typescript, но убрать PropTypes и описать типы пропсов компонентов и аргументов других функций чтобы не использовался тип any
- для всех компонентов должны быть описаны propTypes (для js) или типы/интерфейсы пропсов (для typescript)
- у булок в BurgerConstructor должны быть подписи (верх) и (низ)
- кнопка в BurgerConstrucor по макету должна иметь подпись "оформить заказ"
- получение списка ингредиентов из https://norma.nomoreparties.space/api/ingredients нужно делать один раз в App, сейчас запрос делается 2 раза в BurgerConstructor и в BurgerIngredients
- запрос за данными должен быть обернут в try/catch чтобы обрабатывать ошибки и избегать unhandled rejection в консоли
- переменные которые не переопределяются должны быть объявлены через const (current, isTop, isBottom, isLocked, type, updatedOrderList в BurgerConstructor)
- расчет итоговой стоимости нужно делать в компоненте а не в разметке, чтобы не перемешивать разметку с логикой
- OrderConfirmation должен называться OrderDetails и лежать в отдельном модуле в директории components
- IngredientDetails нужно вынести в отдельный модуль в директории components а не объявлять внутри другого компонента
- OrderDetails и IngredientDetails не должны содержать Modal, а должны содержать только контент попапа, т.к. на следующих этапах будут аналогичные страницы с такой же информацией и появится необходимость переиспользовать компонент
- фоновая подложка модалки (div с классом styles.overlay) должен быть реализован отдельным компонентом ModalOverlay
- в Modal должно быть реализовано закрытие модалки по кнопке esc
- цвет фона модалки по макету должен быть #1C1C21, сейчас используется черный (black)

yuriydyuster committed:
Черновик интерфейса, версия 3.
- Готов основной дизайн
- Захардкожен предварительный заказ
- Работают модальные окна
- Элементы из предваретльного заказа можно уджалят (долбавлять пока нельзя)

yuriydyuster committed:
Исправлены замечания:
 - фоновая подложка модалки (div с классом styles.overlay) должен быть реализован отдельным компонентом ModalOverlay
 - в Modal должно быть реализовано закрытие модалки по кнопке esc
 - цвет фона модалки по макету должен быть #1C1C21, сейчас используется черный (black)
 - AppDashboard - довольно избыточная обертка которая добавляет только один класс, лучше удалить этот компонент и класс добавить непосредственно в App

yuriydyuster committed:
Исправлены замечания:
 - для всех компонентов должны быть описаны типы/интерфейсы пропсов (для typescript)
 - у булок в BurgerConstructor должны быть подписи (верх) и (низ)
 - кнопка в BurgerConstrucor по макету должна иметь подпись "оформить заказ"
 - получение списка ингредиентов из https://norma.nomoreparties.space/api/ingredients нужно делать один раз в App, сейчас запрос делается 2 раза в BurgerConstructor и в BurgerIngredients
 - запрос за данными должен быть обернут в try/catch чтобы обрабатывать ошибки и избегать unhandled rejection в консоли
 - переменные которые не переопределяются должны быть объявлены через const (current, isTop, isBottom, isLocked, type, updatedOrderList в BurgerConstructor)
 - расчет итоговой стоимости нужно делать в компоненте а не в разметке, чтобы не перемешивать разметку с логикой
 - OrderConfirmation должен называться OrderDetails и лежать в отдельном модуле в директории components
 - IngredientDetails нужно вынести в отдельный модуль в директории components а не объявлять внутри другого компонента
 - OrderDetails и IngredientDetails не должны содержать Modal, а должны содержать только контент попапа, т.к. на следующих этапах будут аналогичные страницы с такой же информацией и появится необходимость переиспользовать компонент

andrepolischuk commented on src/components/app/App.tsx:
```diff
@@ -17,14 +18,37 @@ function App() {
+            }
+            setLoadingStatus(false);
+        }
+        try {getIngredients()}
```
нужно исправить: нужно делать await getIngredients иначе try catch не будет отлавливать ошибки асинхронной функции

andrepolischuk commented on src/components/burger-constructor/burger-constructor.tsx:
```diff
@@ -1,66 +1,60 @@
+import OrderDetails from "../order-details/order-details";
 
-function BurgerConstructor (props : any) {
+function BurgerConstructor (props : {
```
можно сделать лучше: стоит типы выносить отдельно, чтобы сделать код более читаемым

```js
type BurgerConstructorProps = {
  ...
}

function BurgerConstructor (props: BurgerConstructorProps) {
  ...
}
```

andrepolischuk commented on src/components/burger-ingredients/burger-ingredients.tsx:
```diff
@@ -73,30 +54,22 @@ const IngredientsTabs = (props: any) => {
+
                                         return (<IngredientCard
-                                            key={index}
+                                            key={index} // @ts-ignore
```
нужно исправил: простите, не заметил на прошлой итерации, в качестве key нужно использовать уникальные постоянные идентификаторы, например _id ингредиента

andrepolischuk commented on src/components/ingredient-card/ingredient-card.tsx:
```diff
@@ -2,73 +2,44 @@ import React, {useState} from 'react';
-    const [count, setCount] = useState(0);
+import IngredientDetails from "../ingredient-details/ingredient-details";
+
+export interface IngredientCardProps  {
```
отлично что тип ингредиента переиспользуется 👍

andrepolischuk commented on src/components/ingredient-details/ingredient-details.tsx:
```diff
@@ -0,0 +1,54 @@
+import styles from "./ingredient-details.module.css";
+
+function IngredientDetails (props: {
+    _id?: string,
```
можно сделать лучше: стоит в IngredientDetails так же использовать тип IngredientCardProps

andrepolischuk commented on src/components/modal/modal.tsx:
```diff
@@ -1,23 +1,45 @@
+
+        return () => document.removeEventListener("keypress", handleEscClose);
+
+    }, []);
```
нужно исправить: при обновлении пропса toClose, нужно чтобы происходило переподписывание на событие. поэтому нужно добавить toClose в зависимости хуков

andrepolischuk reviewed (commented):
добрый вечер, в целом все отлично, внесены практически все правки и даже то что можно было сделать лучше

но еще осталось несколько недочетов
- нужно делать await getIngredients иначе try catch не будет отлавливать ошибки асинхронной функции
- простите, не заметил на прошлой итерации, в качестве key нужно использовать уникальные постоянные идентификаторы, например _id ингредиента
- при обновлении пропса toClose, нужно чтобы происходило переподписывание на событие. поэтому нужно добавить toClose в зависимости хуков

исправьте оставшиеся недоработки и работа будет принята

удачи

yuriydyuster committed:
Исправлены замечания:
 - асинхронный выхов в блоке try/catch
 - переиспользован интерфейс пропсов
 - добавлена зависимость для сайд-эффекта в модальном окне на обновление функции toClose
 - во всех рендераз списков прописаны уникальные ключи для каждого элемента

andrepolischuk reviewed (approved):
отлично, все недочеты исправлены, ваша работа принята 🥳

успехов в дальнейшем обучении

yuriydyuster committed:
Исправлены замечания:
 - скорректированы индексы (key) для React.Fragment в BurgerIngredients
 - скорректированы индексы (key) для рендера массива NutritionFacts в IngredientDetails

 P.S. Урра! Больше никаких варнингов в консоли!!!

yuriydyuster commented:
@andrepolischuk ура!!! Спасибо огромное за ваши комментарии и ревью, было невероятно плезно!!! Я всю голову сломал из-за варнингов в консоли по поводу этих индексов key 🤪
